openfl/federated/__init__.py

```diff
@@ -14,7 +14,11 @@
     from openfl.federated.data import FederatedDataSet  # NOQA
     from openfl.federated.data import KerasDataLoader, TensorFlowDataLoader
     from openfl.federated.task import FederatedModel  # NOQA
-    from openfl.federated.task import KerasTaskRunner, TensorFlowTaskRunner
+    from openfl.federated.task import (
+        KerasTaskRunner,
+        TensorFlowTaskRunner,
+        KerasTensorFlowTaskRunner,
```

Comment:
~Rename all instances of `KerasTensorFlowTaskRunner` to `TensorFlowKerasTaskRunner`. It may feel like a nitpick, but the legacy API namespace is actually `tf.keras`.~

~Current naming may confuse the user to interpret `KerasTensorFlowTaskRunner` as "Tensorflow backend of Keras3"~

Edit: OK, I am confused now. What exactly are we supporting? I thought this was about updating to `KerasTaskRunner` (i.e. the multi-backend Keras3) and `TensorFlowKerasTaskRunner` (i.e. the legacy tf 2.x namespace). Please clarify?

Reply:
- KerasTaskRunner: Keras 2 and TensorFlow 2.13
- TensorFlowTaskRunner: Keras 2 and TensorFlow 2.13
- KerasTensorFlowTaskRunner: Updated to Keras 3 and TensorFlow 2.18

KerasTaskRunner is used by 5 workspaces prior to this PR:

- Keras_cnn_mnist
- Keras_cnn_with_compression
- Keras_nlp
- tf_3dunet_brats
- tf_cnn_histology
- TensorFlowTaskRunner is used by tf_2dunet.

After this PR and https://github.com/securefederatedai/openfl/pull/1204 we will have all 6 workspaces updated with the latest Keras and TensorFlow versions. https://github.com/securefederatedai/openfl/pull/1204 will also move older TensorFlow workspace code to the deprecated folder, which will be removed in the future. Therefore, KerasTaskRunner and TensorFlowTaskRunner will be removed later.

Reply:
I might've missed some discussion. What is the rationale for keeping `KerasTaskRunner` and introducing a new `KerasTensorFlowTaskRunner` instead of just updating `KerasTaskRunner` as a multi-backend Keras3 (as @MasterSkepticista mentioned)

Reply:
Frankly, I remain confused even with this. Here is my proposal:

* `KerasTaskRunner`: A new Keras 3 task runner which is agnostic of any backend that "keras" provides support for, i.e., any keras 3 taskrunner workspace should work with torch/tf/jax/numpy backends.
* `TensorFlowTaskRunner`: The legacy `tf.keras` TaskRunner for `tensorflow<=2.16,>=2.0`. This can remain in maintenance mode for about a year, because community no longer writes entire NNs in pure TensorFlow anymore.
* `PyTorchTaskRunner`: as-is.

Note: This may mean removing support (+code/examples) for `tensorflow==1.x`. This is **completely** safe to do. This API is so old and practically unusable, there is no harm removing it.

Cc: @kta-intel, https://github.com/securefederatedai/openfl/pull/985#issuecomment-2163817246 https://github.com/securefederatedai/openfl/pull/985#issuecomment-2163323526

Reply:
> I might've missed some discussion. What is the rationale for keeping `KerasTaskRunner` and introducing a new `KerasTensorFlowTaskRunner` instead of just updating `KerasTaskRunner` as a multi-backend Keras3 (as @MasterSkepticista mentioned)?
> 
> If we have large enough justification to keep Keras 2 support in this release, we should just add another class to `runner_keras.py`, imo. After this and that other PR you mentioned, it would look like:
> 
> * `runner_keras.KerasTaskRunner` -> supports Keras 3 and TensorFlow 2.18
> * `runner_keras.Keras2TaskRunner` -> supports Keras 2 and TensorFlow 2.13
> * remove/deprecate `TensorFlowTaskRunner`
> 
> `runner_keras.KerasTaskRunner` should support anything that Keras supports (i.e. TF, keras-JAX, keras-torch, etc.), so we won't need a separate `KerasTensorFlowTaskRunner`

I agree, Kevin. One tiny detail though: Keras 2 isn't commonly known, except for us since we closely track tf2 and keras3 naming. Users would only know `tf.keras` and `keras>=3`. So my suggestion (subject to having a consensus) would be to deprecate `TensorFlowTaskRunner` as you mentioned, since it refers to `tf1.x`, but retain the same name for Keras2 (equivalently `tf.keras`).

TL;DR: We deprecate all `tf1.x` code in OpenFL, but retain the class name `TensorFlowTaskRunner` for all `tf2.x` code.

Reply:
> I agree, Kevin. One tiny detail though: Keras 2 isn't commonly known, except for us since we closely track tf2 and keras3 naming. Users would only know `tf.keras` and `keras>=3`. So my suggestion (subject to having a consensus) would be to deprecate `TensorFlowTaskRunner` as you mentioned, since it refers to `tf1.x`, but retain the same name for Keras2 (equivalently `tf.keras`).
> 
> TL;DR: We deprecate all `tf1.x` code in OpenFL, but retain the class name `TensorFlowTaskRunner` for all `tf2.x` code.
> 
> WDYT?

I like this proposal. +1 from me.
Using `TensorFlowTaskRunner` as our vehicle for Keras2 support via `tf.keras` makes a lot more sense

Reply:
TensorFlowTaskRunner  and KerasTensorFlowTaskRunner code removed. KerasTaskRunner is update to support latest versions.